Fix CloudwatchTaskHandler display error

jason810496 opened the pull request:
related: #49470 

## Why

As noted in https://github.com/apache/airflow/pull/49470#discussion_r2239313403, PR #49470 unintentionally broke the display format of `CloudWatchHandler`.

## What

This PR fixes the display issue for `CloudWatchHandler` and ensure it will display properly across different `airflow-core` versions.

jason810496 committed:
Take over ash's fix

jason810496 committed:
Fix datetime serialization error

jason810496 committed:
Fix 'generator is not subscriptable' error

jason810496 committed:
Fix test_cloudwatch_task_handler

ashb reviewed (commented):
This still goes and loads everything into memory for Cloudwatch doesn't it? As we call `io.read()` which does `[x for x in self.stream()]`  -- I wonder if the read code in file_task_handler should check for a `io.stream` function and use that in preference over `io.read`?

That might be a bit complex, but either way I think we need a bit more thought here to not undo all the changes you had for paging/oom of logs

Lee-W commented on airflow-core/src/airflow/utils/log/file_task_handler.py:
```diff
@@ -68,10 +68,10 @@
 # These types are similar, but have distinct names to make processing them less error prone
 LogMessages: TypeAlias = list[str]
-"""The legacy format of log messages before 3.0.2"""
+"""The legacy format of log messages before 3.0.3"""
```
I thought it was 3.0.4

ashb replied:
Yeah, the OOM log PR only got released in 3.0.4

Lee-W commented on providers/amazon/src/airflow/providers/amazon/aws/log/cloudwatch_task_handler.py:
```diff
@@ -72,6 +78,15 @@ def json_serialize(value: Any) -> str | None:
+class DateTimeEncoder(json.JSONEncoder):
+    """Custom JSON encoder to handle datetime serialization."""
+
+    def default(self, obj):
```
```suggestion
    def default(self, obj: object) -> str:
```

Lee-W commented on providers/amazon/src/airflow/providers/amazon/aws/log/cloudwatch_task_handler.py:
```diff
@@ -163,15 +178,32 @@ def upload(self, path: os.PathLike | str, ti: RuntimeTI):
+    def read(self, relative_path, ti: RuntimeTI) -> LegacyLogResponse:
+        messages, logs = self.stream(relative_path, ti)
+
+        return messages, [
```
might be worth making it a function. This list comprehension is not that easy to comprehend

Lee-W commented on providers/amazon/src/airflow/providers/amazon/aws/log/cloudwatch_task_handler.py:
```diff
@@ -275,9 +299,28 @@ def close(self):
+                logs = [(self._event_to_str(event) for event in events)]
+            else:
+                logs = ["\n".join(self._event_to_str(event) for event in events)]
+        except Exception as e:
```
Not sure whehter it's possible to list all possible exception instead of using `Exception`

ashb commented on providers/amazon/src/airflow/providers/amazon/aws/log/cloudwatch_task_handler.py:
```diff
@@ -275,9 +299,28 @@ def close(self):
+        logs: list[str] | list[Generator[str, None, None]]
+        try:
+            events = self.io.get_cloudwatch_logs(stream_name, task_instance)
+            if SUPPORT_STREAM_BASED_READ:
```
Read on CloudwatchTaskHandler should only ever be called for old Airflow Core versions (Newer versions use `CloudwatchRemoteIO` directly instead -- do we need this here?

jason810496 committed:
Fix nits in code review

eladkal commented on providers/amazon/src/airflow/providers/amazon/aws/log/cloudwatch_task_handler.py:
Are the changes to this file backward compatible with Airflow 2.10? PRs that change both core and providers may hide compatibility issue

jason810496 replied:
Yes, and I agree. 
I will test the CloudWatchHandler with Airflow 2.10 as well later on.

jason810496 committed:
Add CloudWatchLogEvent type

jason810496 committed:
Correct return type of .stream method

- .stream method should return gen[str] but it return in gen[dict] in
  previous fix
- Make _parse_cloudwatch_log_event return as str intead of dict can fix
  the problem

jason810496 committed:
Revert change in file_task_handler

jason810496 committed:
Fix test_log_message

jason810496 committed:
Revert file_task_handler change

jason810496 committed:
Fix type annotation for CloudWatchRemoteLogIO

jason810496 reviewed (commented):
The PR _should_ be good to go, I just test it with both Airflow 3 and Airflow 2 with the following setup and both work well.

`breeze start-airflow --backend postgres --integration localstack --db-reset --mount-sources providers-and-tests --use-airflow-version 3.1.3`

<img width="1470" height="956" alt="Image" src="https://github.com/user-attachments/assets/e6eeb518-a884-4505-8b9f-bfcf7f5be037" />

`breeze start-airflow --backend postgres --integration localstack --mount-sources providers-and-tests --use-airflow-version 2.11.0`

<img width="1470" height="956" alt="Image" src="https://github.com/user-attachments/assets/f35e5e5d-cd4e-44e0-b971-006d1c57cdc6" />

Lee-W commented on providers/amazon/src/airflow/providers/amazon/aws/log/cloudwatch_task_handler.py:
```diff
@@ -163,20 +171,35 @@ def upload(self, path: os.PathLike | str, ti: RuntimeTI):
+
+        for group in logs:
+            for msg in group:
+                str_logs.append(f"{msg}\n")
```
```suggestion
        str_logs: list[str] = [
            f"{msg}]n"
            for group in logs
            for msg in group
        ]
```

Lee-W commented on providers/amazon/src/airflow/providers/amazon/aws/log/cloudwatch_task_handler.py:
```diff
@@ -192,29 +215,22 @@ def get_cloudwatch_logs(self, stream_name: str, task_instance: RuntimeTI):
-        return "\n".join(self._event_to_str(event) for event in events)
 
-    def _event_to_dict(self, event: dict) -> dict:
+    def _parse_cloudwatch_log_event(self, event: CloudWatchLogEvent) -> str:
```
```suggestion
    def _parse_log_event_as_dumped_json(self, event: CloudWatchLogEvent) -> str:
```

would this be better this way? i would expect it to be a dict at first glance

Lee-W replied:
or there's a `_event_to_str` in `CloudwatchTaskHandler`

jason810496 replied:
`_parse_log_event_as_dumped_json` is better, thanks!

The purpose of `_event_to_str` and `_parse_log_event_as_dumped_json` are different: 

- `_event_to_str` will parse `CloudWatchLogEvent` typed dict to Airflow 2 compatible format
- `_parse_log_event_as_dumped_json` will only dump the dict as str with proper timestamp

jason810496 committed:
Fix review comments

- consolidate str_logs
- rename _event_to_dict
